src/search/ir_sema_checker.h

```diff
@@ -97,6 +97,42 @@ struct SemaChecker {
       } else {
         v->field->info = &iter->second;
       }
+    } else if (auto v = dynamic_cast<VectorSearchExpr *>(node)) {
+      if (auto iter = current_index->fields.find(v->field->name); iter == current_index->fields.end()) {
+        return {Status::NotOK, fmt::format("field `{}` not found in index `{}`", v->field->name, current_index->name)};
+      } else if (!iter->second.MetadataAs<redis::HnswVectorFieldMetadata>()) {
+        return {Status::NotOK, fmt::format("field `{}` is not a vector field", v->field->name)};
+      } else {
+        v->field->info = &iter->second;
+
+        auto meta = v->field->info->MetadataAs<redis::HnswVectorFieldMetadata>();
+        if (v->vector->values.size() != meta->dim) {
+          return {Status::NotOK,
+                  fmt::format("vector should be of size `{}` for field `{}`", meta->dim, v->field->name)};
+        }
+      }
+    } else if (auto v = dynamic_cast<VectorRangeExpr *>(node)) {
+      if (auto iter = current_index->fields.find(v->field->name); iter == current_index->fields.end()) {
+        return {Status::NotOK, fmt::format("field `{}` not found in index `{}`", v->field->name, current_index->name)};
+      } else if (!iter->second.MetadataAs<redis::HnswVectorFieldMetadata>()) {
+        return {Status::NotOK, fmt::format("field `{}` is not a vector field", v->field->name)};
+      } else {
+        v->field->info = &iter->second;
+
+        auto meta = v->field->info->MetadataAs<redis::HnswVectorFieldMetadata>();
+        if (meta->distance_metric == redis::DistanceMetric::L2 && v->range->val < 0) {
+          return {Status::NotOK, "range cannot be a negative number for l2 distance metric"};
+        }
+
+        if (meta->distance_metric == redis::DistanceMetric::COSINE && (v->range->val < 0 || v->range->val > 2)) {
+          return {Status::NotOK, "range has to be between 0 and 2 for cosine distance metric"};
+        }
+
+        if (v->vector->values.size() != meta->dim) {
+          return {Status::NotOK,
+                  fmt::format("vector should be of size `{}` for field `{}`", meta->dim, v->field->name)};
+        }
+      }
```

Comment:
also in ORDER BY we need to check if the vector literal is provided when the field is typed vector.

Reply:
Double checked this case with `select f4 from ia order by f4 limit 5`. In the case that query only gives a field, it will go through
```
if (auto v = dynamic_cast<SortByClause *>(node)) {
// ...
      if (!iter->second.IsSortable()) {
        return {Status::NotOK, fmt::format("field `{}` is not sortable", v->field->name)};
      } 
    } 
```

So this wouldn't be a problem for vector type in particular, as `SortByClause` will invalidate all non-numeric literal.

In the case that  `ORDER BY vector_field <-> [3...] LIMIT n`, the syntax is guaranteed to be correct by parser.

Reply:
> In the case that `ORDER BY vector_field <-> [3...] LIMIT n`, the syntax is guaranteed to be correct by parser.

Yeah but seems it will be rejected by SemaChecker.

Could you add this example into test cases of sema checker confirm it?

Reply:
Ahh I've read the code in Transformer. 

Seems it will be transformed into `VectorKnnExpr` rather than an `ORDER BY` with a vector literal?

Then I can get the idea, but I have several comments:
- Such a transformation may be better to put into an optimization pass instead of Transformer in the frontend,
- If we really want to put it inside the Transformer, maybe we can delete `vector` member in `SortByClause`.

Reply:
Gotcha! I didn't think of putting it into an optimization pass, but it does make more sense to me; will implement this solution.

Reply:
Removed the transformation in the frontend; will add it in the next PR which will handle everything related to pass for hnsw vector search.